### PR TITLE
feat: Adding Modbus workaround fields to skip Modbus Address errors

### LINKED ISSUE #14671: Telegraf - Input Modbus Plugin crashes Telegraf if it hits an "illegal address"
### Relevant telegraf.conf

```toml
[global_tags]

[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  precision = "1ms"
  #debug = true
  # quiet = false
  # logtarget = "file"
  #logfile = "/var/log/telegraf/debug.log"
  hostname = ""
  omit_hostname = true

[[outputs.file]]
  files = ["/tmp/telegraf/metrics.out"]
  data_format = "influx"

# [[outputs.influxdb_v2]]
#   urls = ["http://XXXXX:9086"]
#   token = "TOKLI"
#   organization = "COMPANY"
#   bucket = "samsung_batteries"
#   # bucket_tag = ""
#   # exclude_bucket_tag = false
#   # timeout = "5s"
#   # http_headers = {"X-Special-Header" = "Special-Value"}
#   # http_proxy = "http://corporate.proxy:3128"
#   # user_agent = "telegraf"
#   # content_encoding = "gzip"
#   ## Enable or disable uint support for writing uints influxdb 2.0.
#   influx_uint_support = true

#   ## Optional TLS Config for use on HTTP connections.
#   # tls_ca = "/etc/telegraf/ca.pem"
#   # tls_cert = "/etc/telegraf/cert.pem"
#   # tls_key = "/etc/telegraf/key.pem"
#   ## Use TLS but skip chain & host verification
#   #insecure_skip_verify = false

[[inputs.modbus]]
  name = "prod"
  timeout = "3s"
  # TCP - connect via Modbus/TCP
  #controller = "tcp://xxxx:602"
  controller = "tcp://xxxxx:602"
  #debug_connection = true
  configuration_type = "request"


#Rack1
[[inputs.modbus.request]]
    slave_id = 1
   # optimization = "max_insert"
   # optimization_max_register_fill = 50
    byte_order = "ABCD"
    register = "input"
    fields = [
# Rack Level
{ measurement = "Battery", name = "Rack Voltage", type = "UINT16", scale=0.1, address = 10200 },
{ measurement = "battery", name = "Mode", type = "UINT16", scale = 1.0, address = 1360 },
        { measurement = "battery", name = "Alarm #1", type = "UINT16", scale = 1.0, address = 1362 },
        { measurement = "battery", name = "Alarm #2", type = "UINT16", scale = 1.0, address = 1363 },
        { measurement = "battery", name = "Alarm #3", type = "UINT16", scale = 1.0, address = 1364 },
        { measurement = "battery", name = "Alarm #4", type = "UINT16", scale = 1.0, address = 1365 },
        { measurement = "battery", name = "Protection #1", type = "UINT16", scale = 1.0, address = 1366 },
        { measurement = "battery", name = "Protection #2", type = "UINT16", scale = 1.0, address = 1367 },
        { measurement = "battery", name = "Protection #3", type = "UINT16", scale = 1.0, address = 1368 },
        { measurement = "battery", name = "Protection #4", type = "UINT16", scale = 1.0, address = 1369 },
        { measurement = "battery", name = "Control Status", type = "UINT16", scale = 1.0, address = 1370 },
        { measurement = "battery", name = "Voltage (String Sensing)", type = "UINT16", scale = 0.1, address = 1373 },
        { measurement = "battery", name = "Voltage (Module Sensing Summation)", type = "UINT16", scale = 0.1, address = 1374 },
        { measurement = "battery", name = "Voltage (Cell Sensing Summation)", type = "UINT16", scale = 0.1, address = 1375 },
        { measurement = "battery", name = "Current (Sensing)", type = "INT16", scale = 0.1, address = 1376 },
        { measurement = "battery", name = "Current (Average for 3 seconds)", type = "INT16", scale = 0.1, address = 1377 },
        { measurement = "battery", name = "SOC (State of Charge)", type = "UINT16", scale = 0.1, address = 1378 },
        { measurement = "battery", name = "SOH (State of Health)", type = "UINT16", scale = 0.1, address = 1379 },
        { measurement = "battery", name = "Current Limit of Charge", type = "UINT16", scale = 0.1, address = 1380 },
        { measurement = "battery", name = "Current Limit of Discharge", type = "UINT16", scale = 0.1, address = 1381 },
        { measurement = "battery", name = "Power Limit of Charge", type = "UINT16", scale = 10.0, address = 1382 },
        { measurement = "battery", name = "Power Limit of Discharge", type = "UINT16", scale = 10.0, address = 1383 },
        { measurement = "battery", name = "Accumulated Amount of Charge Power", type = "UINT16", scale = 1.0, address = 1384 },
        { measurement = "battery", name = "Accumulated Amount of Discharge Power", type = "UINT16", scale = 1.0, address = 1388 },
        { measurement = "battery", name = "Module Power Off Position", type = "UINT16", scale = 1.0, address = 1396 },
        { measurement = "battery", name = "Module Communication Fault Position", type = "UINT16", scale = 1.0, address = 1397 },
        { measurement = "battery", name = "Cell Voltage (Max Difference)", type = "UINT16", scale = 0.001, address = 1398 },
        { measurement = "battery", name = "Cell Voltage (Average)", type = "UINT16", scale = 0.001, address = 1399 },
        { measurement = "battery", name = "Cell Voltage (Max(1st))", type = "UINT16", scale = 0.001, address = 1400 },
        { measurement = "battery", name = "Cell Voltage (Max(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1401 },
        { measurement = "battery", name = "Cell Voltage (Max(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1402 },
        { measurement = "battery", name = "Cell Voltage (Min(1st))", type = "UINT16", scale = 0.001, address = 1403 },
        { measurement = "battery", name = "Cell Voltage (Min(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1404 },
        { measurement = "battery", name = "Cell Voltage (Min(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1405 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd))", type = "UINT16", scale = 0.001, address = 1406 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1407 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1408 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd))", type = "UINT16", scale = 0.001, address = 1409 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1410 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1411 },
        { measurement = "battery", name = "Cell Temperature (Max Difference)", type = "UINT16", scale = 0.01, address = 1412 },
        { measurement = "battery", name = "Cell Temperature (Average)", type = "INT16", scale = 0.01, address = 1413 },
        { measurement = "battery", name = "Cell Temperature (Max(1st))", type = "INT16", scale = 0.01, address = 1414 },
        { measurement = "battery", name = "Cell Temperature (Max(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1415 },
        { measurement = "battery", name = "Cell Temperature (Max(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1416 },
        { measurement = "battery", name = "Cell Temperature (Min(1st))", type = "INT16", scale = 0.01, address = 1417 },
        { measurement = "battery", name = "Cell Temperature (Min(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1418 },
        { measurement = "battery", name = "Cell Temperature (Min(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1419 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd))", type = "INT16", scale = 0.01, address = 1420 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1421 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1422 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd))", type = "INT16", scale = 0.01, address = 1423 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1424 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1425 },
        { measurement = "battery", name = "Module PCB Temperature (Max)", type = "INT16", scale = 0.01, address = 1426 },
        { measurement = "battery", name = "Module PCB Temperature (Max) Module Position", type = "UINT16", scale = 1.0, address = 1427 },
        { measurement = "battery", name = "Module PCB Temperature (Min)", type = "INT16", scale = 0.01, address = 1428 },
        { measurement = "battery", name = "Module PCB Temperature (Min) Module Position", type = "UINT16", scale = 1.0, address = 1429 },
        { measurement = "battery", name = "Work Cell Index", type = "UINT16", scale = 1.0, address = 1434 },
        { measurement = "battery", name = "Sensor Status Register", type = "UINT16", scale = 1.0, address = 1435 },
        { measurement = "battery", name = "Module Status Register (Scan)", type = "UINT16", scale = 1.0, address = 1436 },
        { measurement = "battery", name = "Module Status Register (Balancing)", type = "UINT16", scale = 1.0, address = 1437 },
        { measurement = "battery", name = "Module Status Register (Address)", type = "UINT16", scale = 1.0, address = 1438 },
        { measurement = "battery", name = "Module Status Register (Voltage)", type = "UINT16", scale = 1.0, address = 1439 },
        { measurement = "battery", name = "Module Status Register (Temperature)", type = "UINT16", scale = 1.0, address = 1440 },
        { measurement = "battery", name = "Difference Voltage (Max(Cell Sensing Summation))", type = "UINT16", scale = 0.01, address = 1441 },
        { measurement = "battery", name = "Difference Voltage (Max(Module Sensing))", type = "UINT16", scale = 0.01, address = 1442 },
        { measurement = "battery", name = "Difference Voltage (Max) Module Position", type = "UINT16", scale = 1.0, address = 1443 },
        { measurement = "battery", name = "Difference Voltage (Min(Cell Sensing Summation))", type = "UINT16", scale = 0.01, address = 1444 },
        { measurement = "battery", name = "Difference Voltage (Min(Module Sensing))", type = "UINT16", scale = 0.01, address = 1445 },
        { measurement = "battery", name = "Difference Voltage (Min) Module Position", type = "UINT16", scale = 1.0, address = 1446 },
        { measurement = "battery", name = "State", type = "UINT16", scale = 1.0, address = 1447 },
        { measurement = "battery", name = "Cell V Distribution (Under 20mV)", type = "UINT16", scale = 1.0, address = 1448 },
        { measurement = "battery", name = "Cell V Distribution (Under V-IMB Alarm Level)", type = "UINT16", scale = 1.0, address = 1449 },
        { measurement = "battery", name = "Cell V Distribution (Under V-IMB Protection Level)", type = "UINT16", scale = 1.0, address = 1450 },
        { measurement = "battery", name = "Cell V Distribution (V-IMB Protection Level or More)", type = "UINT16", scale = 1.0, address = 1451 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max)", type = "UINT16", scale = 1.0, address = 1452 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max) Module Position", type = "UINT16", scale = 1.0, address = 1453 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max) Cell Position", type = "UINT16", scale = 1.0, address = 1454 },
        { measurement = "battery", name = "Max Balancing Time (minute)", type = "UINT16", scale = 1.0, address = 1455 },
        { measurement = "battery", name = "Current (Sensing_Sub)", type = "INT16", scale = 0.1, address = 1457 },
        { measurement = "battery", name = "Min Cell Deviation Index", type = "INT16", scale = 1.0, address = 1458 },
        { measurement = "battery", name = "Min Cell Deviation Index Module Position", type = "UINT16", scale = 1.0, address = 1459 },
        { measurement = "battery", name = "Min Cell Deviation Index Cell Position", type = "UINT16", scale = 1.0, address = 1460 },
        { measurement = "battery", name = "SOC (State of Charge(Ref))", type = "INT16", scale = 0.1, address = 1461 },
        { measurement = "battery", name = "Module Ambient Temperature (Max)", type = "INT16", scale = 0.01, address = 1466 },
        { measurement = "battery", name = "Module Ambient Temperature (Max) Module Position", type = "UINT16", scale = 1.0, address = 1467 },
        { measurement = "battery", name = "Module Ambient Temperature (Max) Sensor Position", type = "UINT16", scale = 1.0, address = 1468 },
        { measurement = "battery", name = "Module Ambient Temperature (Min)", type = "INT16", scale = 0.01, address = 1469 },
        { measurement = "battery", name = "Module Ambient Temperature (Min) Module Position", type = "UINT16", scale = 1.0, address = 1470 },
        { measurement = "battery", name = "Module Ambient Temperature (Min) Sensor Position", type = "UINT16", scale = 1.0, address = 1471 },
        { measurement = "battery", name = "Module PCB Temperature (Max) Sensor Position", type = "UINT16", scale = 1.0, address = 1472 },
        { measurement = "battery", name = "Module PCB Temperature (Min) Sensor Position", type = "UINT16", scale = 1.0, address = 1473 },

]
```

### Logs from Telegraf

```text
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1373[12]...
2024-02-02T10:59:51Z D! [inputs.modbus] got input@1373[12]: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 100 0 5 0 48]
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (String Sensing) with offset 0 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (Module Sensing Summation) with offset 2 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (Cell Sensing Summation) with offset 4 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current (Sensing) with offset 6 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current (Average for 3 seconds) with offset 8 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field SOC (State of Charge) with offset 10 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field SOH (State of Health) with offset 12 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current Limit of Charge with offset 14 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current Limit of Discharge with offset 16 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Power Limit of Charge with offset 18 with len 2: [0 100] --> 1000
2024-02-02T10:59:51Z D! [inputs.modbus]   field Power Limit of Discharge with offset 20 with len 2: [0 5] --> 50
2024-02-02T10:59:51Z D! [inputs.modbus]   field Accumulated Amount of Charge Power with offset 22 with len 2: [0 48] --> 48
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1388[1]...
2024-02-02T10:59:51Z D! [inputs.modbus] got input@1388[1]: [48 48]
2024-02-02T10:59:51Z D! [inputs.modbus]   field Accumulated Amount of Discharge Power with offset 0 with len 2: [48 48] --> 12336
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1396[34]...
2024-02-02T10:59:51Z E! [inputs.modbus] Error in plugin: slave 1: modbus: exception '2' (illegal data address), function '132'
2024-02-02T10:59:51Z D! [inputs.modbus] Reconnecting to tcp://10.0.32.11:602...
2024-02-02T10:59:51Z D! [agent] Stopping service inputs
2024-02-02T10:59:51Z D! [agent] Input channel closed
2024-02-02T10:59:51Z D! [agent] Stopped Successfully
2024-02-02T10:59:51Z E! [telegraf] Error running agent: input plugins recorded 1 errors
```

### System info

Telegraf 1.29.4, Linux Native, Mac Native, Docker, doesn't make a difference :) 

### Docker

_No response_

### Steps to reproduce

1. Generate your telegraf.conf config. 
2. Try to debug and check your work.
3. Notice you run into an exception, "illegal address"
4. Do the same registers with simple python code, the code does NOT run in an expeption and stops. 
Here is an example script:
`from pyModbusTCP.client import ModbusClient

 

client = ModbusClient('xxxx', 602, auto_open=True)

client.debug = True

register = [1395]

 

for i in register:

    print(client.read_input_registers(i))

 

 

print(client.read_input_registers(1395))  `

### Expected behavior

Telegraf does not crash.

I would want it still runs and continue on. 
It would be cool if Telegraf could add information (like a tag or field) with affected (non working) fields so that you can easily see what addresses are failing.

Also it would be great if you could maybe set per configuration, what data should go in the field which is not able to gather information from.

### Actual behavior

It crashes and I don't get ANY data - not even from the hundreds of registers which work :) 

In reality, it could be that 1 from 11 racks is defective and needs to be turned off.
I would get the error now - and don't get any data from 10 racks anymore until I change the configuration to work around the currently defective 1 rack.

Hence I think this should be handled differently/better :-) 

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by spkis**
## Summary

This change has been made to accomplish two elements. 
1. Other Telegraf Plugins, like CSV has "Skip Errors". I think from a design perspective, it would be good if all adapters have a skip error flag.
2. I agree that generally, I would like to know if there is an issue with a modbus register. However, I would want to know this in debug/testing/development phase only. 
3. In Production systems, I rather skip errors and still import data instead of having no data at all because of an error. 

To maybe give you a bit more context:
I am using this modbus input plugin to monitor batteries from Battery Energy Storage Systems. Those have multiple racks of batteries per technical unit. 
If we update now racks to get a new firmware and one rack is not able to get a new firmware, it will not be able to respond with the registers. 
However, the telegraf integration would already harvest all other racks successfully.... 
Without the fix, it will now harvest nothing. 

We cannot make multiple telegraf processes - half of the battery management systems are simply too old and only allow two clients to read information from.
The other batteries support multiple reads, however we require additional physical hardware for that. 

This fix allows us to utilize telegraf even for those "funky" scenarios. Hence also added it as workaround :-) 

## Checklist

- [X] No AI generated code was used in this PR

## Related issues

Telegraf - Input Modbus Plugin crashes Telegraf if it hits an "illegal address" #14671

resolves #14671

### CODE REVIEW

**Comment by spkis** [comment]
!signed-cla

**Comment by powersj** [comment]
@spkis,

I have read through the issue you filed and now this PR. I want to level-set our understanding of the current behavior:

Telegraf and the python client you are using operate and behavior in different ways. I do not believe we need to have the same behavior as this python client either. If you want to use the python client and telegraf, then feel free to and have it pass metrics to Telegraf via the `exec` plugin.

In telegraf, we group requests together to be more efficient in collecting data.
If anything in that slave or group fails, then an error is produced. This behavior is not a crash. Telegraf continues to operate and does not produce a trace due to a catastrophic failure.

Other plugins now have connection retry logic, which does not apply here.

While I can understand the desire to continue collected data, you would still be missing out on data from this slave, specifically from the batch that failed. This would introduce unexpected, undefined behavior for our users. In that, some chucks that have errors would be omitted and others could still produce data.

For the reasons above, I am not inclined to make this change.

**Comment by spkis** [comment]
@powersj 
I have configured multiple input plugins in my configuration file. If now only one of those fails, the complete telegraf process doesn't start. 
I am not able to configure - due to hardware limitations - one separate telegraf process per Rack but can only make one telegraf process to query each rack against one BMS. 

The problem is that a handful of registers - due to wrong firmware of the rack - are not the same/working like all others. 
We are not able to update themselves the firmware - and currently we don't get support from the vendor either. 

With this being the case, I would now need to maintain up to 20 different telegraf configurations, because one or two registers are different/not working, since there is no "ignore/skip error" of modbus addresses. 

Additionally, I am not 100% how the BMS reacts if one rack is offline. Will it also give back a invalid address? I am not sure, that I could even test that tbh.
But if one rack fail will also cause modbus address errors returned, then the complete telegraf process would stop working again. 

But in this scenario, again NO data at all would be collected, since telegraf modbus input plugin does only start with all modbus address working. and I believe that it would stop working, as soon as it hits ONE failed address. 
Are you saying, that telegraf would continue to run IF there would suddenly be an address error? Because I simply cannot confirm this and would ask, if you have actually tested this behavior.

I don't want to switch to python or node-red however, with the current feature scope as is, this would be the consequence, since I cannot maintain 5000 possible telegraf variations (while only like two or three registers each time are simply reserverd/not working or working).

Basically, if only one defined input plugin from a slave would fail, and the rest would simply work and continue to run, that would be already fine for me and basically what we want to achieve. If the data from this batch even would be missed, also fine, though I think it should only skip the one field which it cannot pick up. 

Since this is adding a specific new modbus_workaround_flag, why is this introducing breaking changes or why is this fine for other plugins and not that one?

**Comment by baxtheman** [comment]
With multiple modbus devices enviroment, the issue raised by @spkis is real.
It's important to a have  more resilient plugin.

**Comment by powersj** [comment]
> If now only one of those fails, the complete telegraf process doesn't start.

This is expected. How does Telegraf know that your config is not invalid? For example, maybe you typo'ed an address or have the wrong address or password. We have seen from users that if Telegraf starts running, they assume that everything is working as expected.

> since there is no "ignore/skip error" of modbus addresses.

If you want to skip the entire plugin, then that would be accepted. However, taking in only partial bits of the data will not. As I already mentioned this leads to very unexpected behaivor where some data is always skipped, which may be completely valid.

**Comment by baxtheman** [comment]
@powersj 
Yes, I understand your point about the initial validation for all plugins in bootstrap phase.

I ask @spkis if having a proxy in the middle between telegraf and all modbus devices can solve the situation:
https://github.com/tiagocoutinho/modbus-proxy
The proxy permits the connections with the plugins also if modbus device is temporary down.

**Comment by spkis** [comment]
@powersj 
Can you tell me, why this behavior of just skipping errors is fine with the input plugins for CSV, or other file-based plugins, and not OK with modbus. 
I think this should be a general design pattern in Telegraf, to allow skip errors. 

Also I am fine with initial validation, but to your point: If Telegraf runs and I lose a few devices, it will not stop but continue on?
If not - this is not industrial "fail proof" enough in the current form. 

@baxtheman 
Well, if I do always deploy the proxy WITH telegraf as a docker compose - in theory yes. 
However, I rather have a native solution without the need of 3rd party apps or proxies.

I still think this should be handled different and better by telegraf.

**Comment by powersj** [comment]
> just skipping errors is fine with the input plugins for CSV, or other file-based plugins, and not OK with modbus.

You are comparing parser errors with plugin errors. They are not the something and will have different behavior.

Do you plan to update this PR?

**Comment by spkis** [comment]
I did testing now - the modbus proxy works for losing connection to devices, but it will also stop executing if it hits an address error. 
I currently cannot test, if just one rack goes offline instead of the complete BMS system. 

Accepting the PR would also be wrong - I agree.

However, what other possibilities do we have here? Or is "this it" and there is no way or willingness to update telegraf to support such "edge" cases.

**Comment by spkis** [comment]
EDIT: I did research internally. Apparently, we get the batteries already delivered with different firmware images on the racks.
I was able to reverse engineer the configuration files of the devices and luckily, we have some new modbus registers which are relevant for warranty in newer firmware versions.
Hence we are trying to get the firmware updates again with more pressure now.
=> The different versions are not from production updates, hence the issue I tried to workaround, isn't needed and I will try to workaround those racks upfront.

I will still utilize the modbus proxy - for some scenarios it looks like a helpful tool Thanks again for the hint @baxtheman

**Comment by baxtheman** [comment]
@spkis 
Welcome, I am still investigating for alternatives modbus proxy because that github prj seems not battle-tested.
Maybe a tcp proxy based on nginx or haproxy will be better.
